🐛 Only refresh bootstrap token if needed, requeue in all cases where node hasn't joined yet

AndiDog opened the pull request:
<!-- Thanks for sending a pull request! Here are some tips for you:
    1. If this is your first time, please read our contributor guidelines: https://github.com/kubernetes-sigs/cluster-api/blob/main/CONTRIBUTING.md#contributing-a-patch and developer guide https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/book/src/developer/guide.md

    2. Please add an icon to the title of this PR (see https://sigs.k8s.io/cluster-api/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones
    the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) 
-->

**What this PR does / why we need it**:

Right now, the bootstrap token is refreshed every time until nodes have joined. This creates unnecessary etcd writes. Only extend the expiry timestamp (called "refresh" in code) if we're getting a bit closer to reaching it. Particularly if KubeadmConfig reconciliation is triggered from other sources than CAPA's `RequeueAfter` interval, this can pile up to quite noisy logs and API accesses.

Along the way, I did some other improvements and will add inline comments to explain.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

This was a drive-by along the way of fixing https://github.com/giantswarm/roadmap/issues/2217, but isn't really related.

<!-- 
Please label this pull request according to what area(s) you are addressing. For reference on PR/issue labels, see: https://github.com/kubernetes-sigs/cluster-api/labels?q=area+

Area example:
/area runtime-sdk
-->

/area provider/bootstrap-kubeadm
/area testing

AndiDog commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go:
```diff
@@ -321,12 +323,40 @@ func (r *KubeadmConfigReconciler) refreshBootstrapToken(ctx context.Context, con
 	}
 
-	log.Info("Refreshing token until the infrastructure has a chance to consume it")
-	if err := refreshToken(ctx, remoteClient, token, r.TokenTTL); err != nil {
```
Moving code out of `refreshToken` because it's only used once and I didn't see the point of having two functions called "refresh token if necessary".

AndiDog commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go:
```diff
@@ -321,12 +323,40 @@ func (r *KubeadmConfigReconciler) refreshBootstrapToken(ctx context.Context, con
+			return ctrl.Result{}, errors.Wrapf(err, "can't parse expiration time of bootstrap token")
+		}
+
+		if expiration.After(time.Now().UTC().Add(r.TokenTTL * 5 / 6)) {
```
Refresh if 1/6-th of the TTL has passed. We can still change this, but since `RequeueAfter` is now consistently `r.TokenTTL / 3` (which I consider a sane value even for intermittent errors), the next regular reconciliation would refresh.

CecileRobertMichon replied:
Let's add a code comment here that explains this

Since this value needs to be tightly coupled with `tokenCheckRefreshOrRotationInterval`, we might want to either define them in the same place in the code or make them depend on each other so we don't inadvertently change one but not the other in the future.

AndiDog commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go:
```diff
@@ -321,12 +323,40 @@ func (r *KubeadmConfigReconciler) refreshBootstrapToken(ctx context.Context, con
 	}
 	return ctrl.Result{
-		RequeueAfter: r.TokenTTL / 2,
+		RequeueAfter: r.TokenTTL / 3,
```
Consistency. Also, if we only reconcile twice within the TTL, the first reconciliation at halftime (TTL / 2) must not fail. Better give more chances if there are intermittent errors.

AndiDog commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go:
```diff
@@ -554,6 +584,9 @@ func (r *KubeadmConfigReconciler) joinWorker(ctx context.Context, scope *Scope)
 	if res, err := r.reconcileDiscovery(ctx, scope.Cluster, scope.Config, certificates); err != nil {
 		return ctrl.Result{}, err
 	} else if !res.IsZero() {
+		if res.RequeueAfter == 0 {
```
This and the below code line change were IMO missing cases where the reconciler returned that no requeueing is needed, but we need that while nodes haven't joined since the token may expire. It's covered in a changed test, so please double-check if that makes sense.

fabriziopandini replied:
I don't think this change is necessary.
reconcileDiscovery takes care of creating the token, it is not linked to the refresh token process

AndiDog commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller_test.go:
```diff
@@ -508,7 +508,7 @@ func TestKubeadmConfigReconciler_Reconcile_GenerateCloudConfigData(t *testing.T)
 		Client:              myclient,
 		SecretCachingClient: myclient,
-		Tracker:             remote.NewTestClusterCacheTracker(logr.New(log.NullLogSink{}), myclient, myclient.Scheme(), client.ObjectKey{Name: cluster.Name, Namespace: cluster.Namespace}),
+		Tracker:             remote.NewTestClusterCacheTracker(logr.New(log.NullLogSink{}), myclient, myclient, myclient.Scheme(), client.ObjectKey{Name: cluster.Name, Namespace: cluster.Namespace}),
```
I added a new parameter to `NewTestClusterCacheTracker`. Generally, I don't like how all unit tests mix the management cluster's k8s client with the workload cluster's client ("remote") and hence store all objects within the same fakeClient's storage. This can be confusing for developers since you don't really know whether to expect something in the management or workload cluster. So for the test I added, there's now a split between those two types of client → bootstrap data in the management cluster, bootstrap token secret in the workload cluster. With this, we also test that CAPA actually puts objects in the correct cluster. However, I didn't adapt all tests because that would have blown up the PR scope.

sbueringer commented:
/ok-to-test

AndiDog commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller_test.go:
```diff
@@ -1070,18 +1070,52 @@ func TestBootstrapTokenTTLExtension(t *testing.T) {
+	g.Expect(l.Items).To(HaveLen(2)) // control plane vs. worker
 
-	// ensure that the token is refreshed...
+	t.Log("Ensure that the token secret is not updated while it's still fresh")
```
Test log is better than a code comment since you can then see which controller log lines belonged to which part of the test, and these logs are only displayed on test failure (except for `go test -v`).

AndiDog commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller_test.go:
```diff
@@ -1070,18 +1070,52 @@ func TestBootstrapTokenTTLExtension(t *testing.T) {
 		tokenExpires[i] = item.Data[bootstrapapi.BootstrapTokenExpirationKey]
 	}
 
-	<-time.After(1 * time.Second)
```
Instead of having tests sleep, which is a bad developer experience, I'm now simulating close-to-expired tokens.

AndiDog commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller_test.go:
```diff
@@ -1070,18 +1070,52 @@ func TestBootstrapTokenTTLExtension(t *testing.T) {
+	g.Expect(remoteClient.List(ctx, l, client.ListOption(client.InNamespace(metav1.NamespaceSystem)))).To(Succeed())
+	g.Expect(l.Items).To(HaveLen(2))
+
+	for i, item := range l.Items {
```
This part is a new subtest

AndiDog commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller_test.go:
```diff
@@ -1099,20 +1133,28 @@ func TestBootstrapTokenTTLExtension(t *testing.T) {
 		result, err := k.Reconcile(ctx, req)
 		g.Expect(err).ToNot(HaveOccurred())
-		g.Expect(result.RequeueAfter).NotTo(BeNumerically(">=", k.TokenTTL))
+		g.Expect(result.RequeueAfter).To(Equal(k.TokenTTL / 3))
```
I consider `TTL / 3` much more than just an implementation detail by now and therefore test for the exact value. It's important that the `RequeueAfter` interval leads to multiple reconciliation attempts before the TTL expires, so testing for `0 < result.RequeueAfter < TokenTTL` isn't sufficient.

AndiDog commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller_test.go:
```diff
@@ -1230,12 +1275,13 @@ func TestBootstrapTokenRotationMachinePool(t *testing.T) {
 
 	objects = append(objects, createSecrets(t, cluster, initConfig)...)
 	myclient := fake.NewClientBuilder().WithObjects(objects...).WithStatusSubresource(&bootstrapv1.KubeadmConfig{}, &expv1.MachinePool{}).Build()
+	remoteClient := fake.NewClientBuilder().Build()
```
See above comment for why I distinguish management vs. workload cluster client

fabriziopandini commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go:
```diff
@@ -634,7 +667,7 @@ func (r *KubeadmConfigReconciler) joinWorker(ctx context.Context, scope *Scope)
 		return ctrl.Result{}, err
 	}
-	return ctrl.Result{}, nil
+	return ctrl.Result{RequeueAfter: r.TokenTTL / 3}, nil
```
if we are changing this here, please add a comment explaining why we are requeing, eg.
// joinWorker completed, ensuring to reconcile this object again so we keep refreshing the bootstrap token until it is consumed

Also the same change should be applied to joinControlPlane

enxebre commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go:
```diff
@@ -739,7 +771,8 @@ func (r *KubeadmConfigReconciler) joinControlplane(ctx context.Context, scope *S
 
-	return ctrl.Result{}, nil
+	// Ensure reconciling this object again so we keep refreshing the bootstrap token until it is consumed
+	return ctrl.Result{RequeueAfter: r.TokenTTL / 3}, nil
```
nit: given this is using r.TokenTTL / 3 in multiple times I'd put it in a var e.g tokenRequeueAfter that is invoked everywhere

AndiDog replied:
done

fabriziopandini replied:
This seems to be still pending

AndiDog commented:
This API diff was detected as test error:

```
NewTestClusterCacheTracker: changed from func(github.com/go-logr/logr.Logger, sigs.k8s.io/controller-runtime/pkg/client.Client, *k8s.io/apimachinery/pkg/runtime.Scheme, k8s.io/apimachinery/pkg/types.NamespacedName, ...string) *ClusterCacheTracker to func(github.com/go-logr/logr.Logger, sigs.k8s.io/controller-runtime/pkg/client.Client, sigs.k8s.io/controller-runtime/pkg/client.Client, *k8s.io/apimachinery/pkg/runtime.Scheme, k8s.io/apimachinery/pkg/types.NamespacedName, ...string) *ClusterCacheTracker
```

What do I need to do here? I changed the function signature on purpose.

AndiDog commented:
@enxebre @fabriziopandini The requested changes are included. We found in [chat](https://kubernetes.slack.com/archives/C8TSNPY4T/p1693212998183259?thread_ts=1693212367.802049&cid=C8TSNPY4T) that the apidiff test error is not mandatory, so I think we should go ahead since only a test function signature was changed. If that change makes sense, please have a look since I think this PR is ready now.

fabriziopandini commented:
changes looks good to me, but it would be great if someone else give a look at this

/test pull-cluster-api-e2e-full-main

AndiDog commented:
> changes looks good to me, but it would be great if someone else give a look at this

@enxebre Would you like to have a closer look?

AndiDog commented:
@enxebre @JoelSpeed could you review this one? Thank you.

neolit123 commented:
security people might not agree with keeping bootstrap tokens for longer periods of time
https://github.com/kubernetes/kubernetes/issues/119639

CecileRobertMichon commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go:
```diff
@@ -321,12 +323,40 @@ func (r *KubeadmConfigReconciler) refreshBootstrapToken(ctx context.Context, con
+
+		if expiration.After(time.Now().UTC().Add(r.TokenTTL * 5 / 6)) {
+			// We still have lots of time until it expires
+			log.Info("Token needs no refresh")
```
would it make sense to log how much time is left before expiration, maybe as a higher verbosity for debugging purposes?

CecileRobertMichon reviewed (commented):
lgtm overall, a few minor comments

AndiDog commented:
> security people might not agree with keeping bootstrap tokens for longer periods of time [kubernetes/kubernetes#119639](https://github.com/kubernetes/kubernetes/issues/119639)
> 
> if CAPI is keeping them as a maximum of 24h, sgtm.

We have refresh (token does not expire while the node didn't join) vs. rotation (machine pools only – keep creating a new token so that new nodes can start up at any time). There's no time limit for the refresh case, though. If we think this is problematic, we should tackle it in a separate issue. My PR only fixes the constant refreshing (= changing `expiration` value on the `Secret`) on every reconciliation, which is really just a minor bug and not changing the security stance, I think.

AndiDog reviewed (commented):
I addressed the review comments and also tested the change manually.

CecileRobertMichon reviewed (commented):
/lgtm
/assign @fabriziopandini @enxebre

vincepri commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go:
```diff
@@ -321,12 +323,42 @@ func (r *KubeadmConfigReconciler) refreshBootstrapToken(ctx context.Context, con
+
+	secretExpiration := bootstrapsecretutil.GetData(secret, bootstrapapi.BootstrapTokenExpirationKey)
+	if secretExpiration == "" {
+		log.Info(fmt.Sprintf("Token has no valid value for %s, writing new expiration timestamp", bootstrapapi.BootstrapTokenExpirationKey))
```
In which case would the expiration be empty?

vincepri commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go:
```diff
@@ -321,12 +323,42 @@ func (r *KubeadmConfigReconciler) refreshBootstrapToken(ctx context.Context, con
+	newExpiration := time.Now().UTC().Add(r.TokenTTL).Format(time.RFC3339)
+	secret.Data[bootstrapapi.BootstrapTokenExpirationKey] = []byte(newExpiration)
+	log.Info("Refreshing token until the infrastructure has a chance to consume it", "oldExpiration", secretExpiration, "newExpiration", newExpiration)
+	err = remoteClient.Update(ctx, secret)
+	if err != nil {
```
```suggestion
	if err := remoteClient.Update(ctx, secret); err != nil {
```

vincepri commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go:
```diff
@@ -321,12 +323,42 @@ func (r *KubeadmConfigReconciler) refreshBootstrapToken(ctx context.Context, con
+	// Extend TTL for existing token
+	newExpiration := time.Now().UTC().Add(r.TokenTTL).Format(time.RFC3339)
+	secret.Data[bootstrapapi.BootstrapTokenExpirationKey] = []byte(newExpiration)
+	log.Info("Refreshing token until the infrastructure has a chance to consume it", "oldExpiration", secretExpiration, "newExpiration", newExpiration)
```
```suggestion
	log.Info("Bootstrap data hasn't yet been consumed by an infrastructure provider; refreshing token", "oldExpiration", secretExpiration, "newExpiration", newExpiration)
```

vincepri commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go:
```diff
@@ -321,12 +323,42 @@ func (r *KubeadmConfigReconciler) refreshBootstrapToken(ctx context.Context, con
+		skipTokenRefreshIfExpiringAfter := now.Add(r.skipTokenRefreshIfExpiringAfter())
+		if expiration.After(skipTokenRefreshIfExpiringAfter) {
+			// We still have lots of time until it expires
+			log.V(2).Info("Token needs no refresh", "tokenExpiresInSeconds", expiration.Sub(now).Seconds())
```
Is this log line needed?

sbueringer replied:
2 is the recommended production log level. Let's use 3

vincepri commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go:
```diff
@@ -321,12 +323,42 @@ func (r *KubeadmConfigReconciler) refreshBootstrapToken(ctx context.Context, con
+		now := time.Now().UTC()
+		skipTokenRefreshIfExpiringAfter := now.Add(r.skipTokenRefreshIfExpiringAfter())
+		if expiration.After(skipTokenRefreshIfExpiringAfter) {
+			// We still have lots of time until it expires
```
This comment says we have "lots of time", although the check above only checks that the expiration is within the window

vincepri commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go:
```diff
@@ -819,6 +854,27 @@ func (r *KubeadmConfigReconciler) resolveSecretPasswordContent(ctx context.Conte
 	return data, nil
 }
 
+// skipTokenRefreshIfExpiringAfter returns a duration from "now". If the token's expiry timestamp is after
+// `now + skipTokenRefreshIfExpiringAfter()`, it does not yet need a refresh.
+func (r *KubeadmConfigReconciler) skipTokenRefreshIfExpiringAfter() time.Duration {
```
Was the name/comment a leftover? The function just returns a duration, rather than a duration from `now`

AndiDog commented:
@vincepri Please have a look. I've pushed merge+changes so you can still see in GitHub what exactly has changed (hopefully). I can squash to a single commit once all looks okay.

vincepri commented:
ok good to squash for me!

vincepri reviewed (commented):
LGTM
@sbueringer do you want to give it a pass?

sbueringer commented:
Yup would be good. I hope I get around to it in the next few days

sbueringer commented on bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go:
```diff
@@ -319,12 +321,42 @@ func (r *KubeadmConfigReconciler) refreshBootstrapToken(ctx context.Context, con
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to get bootstrap token secret in order to refresh it")
+	}
+	log = log.WithValues("secretName", secret.Name)
```
```suggestion
	log = log.WithValues("Secret", klog.KObj(secret))
```

sbueringer reviewed (commented):
Looks good to me.

But not exactly an expert on this part of our code

/assign @fabriziopandini

AndiDog commented:
@sbueringer I applied both suggestions.

sbueringer commented:
Thx!

/lgtm

/assign @fabriziopandini

sbueringer commented:
/test pull-cluster-api-e2e-main

vincepri reviewed (approved):
/approve
